server/etcdserver/api/v3rpc/health.go

```diff
@@ -18,37 +18,57 @@ import (
 	"go.uber.org/zap"
 	"google.golang.org/grpc/health"
 	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
+	"go.etcd.io/etcd/server/v3/etcdserver"
 )
 
 const (
 	allGRPCServices = ""
 )
 
-type HealthNotifier interface {
-	StartServe()
-	StopServe(reason string)
+type etcdServerHealth interface {
+	StopNotify() <-chan struct{}
 }
 
-func NewHealthNotifier(hs *health.Server, lg *zap.Logger) HealthNotifier {
+func NewHealthNotifier(hs *health.Server, s *etcdserver.EtcdServer, defragNotifier <-chan defragStatus) {
 	if hs == nil {
 		panic("unexpected nil gRPC health server")
 	}
-	if lg == nil {
-		lg = zap.NewNop()
-	}
-	hc := &healthChecker{hs: hs, lg: lg}
+	hc := &healthNotifier{hs: hs, s: s, lg: s.Logger(), defragNotifier: defragNotifier, stopGRPCServiceOnDefrag: s.Cfg.ExperimentalStopGRPCServiceOnDefrag}
 	// set grpc health server as serving status blindly since
 	// the grpc server will serve iff s.ReadyNotify() is closed.
-	hc.StartServe()
-	return hc
+	hc.startServe()
+	go hc.watchDefragStatus()
```

Comment:
I'm worried about livecycle of this gorutine. I would recommend not using goroutines here, but a simple variable with mutex.

Reply:
okay. This will be more simple and safer.